-1

**Failed Tests**:   Build 
**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-09dd72/46342/summary.html
**COMMIT**: 3bd2dfc54cb2bdfa0a93bdb19d4c7cad0743888a
**CMSSW**: CMSSW_15_1_X_2025-05-23-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48155/46342/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling alpaka/cuda edm plugin src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitConfigurationESProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-05-23-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-05-23-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_15_1_X_2025-05-23-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/dd4hep/v01-29-00-a6251f4ae4aef60eab61c637c5fb989d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/alpaka/1.2.0-9ea5b7b50bc68454be5c438069f1aa33/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/boost/1.80.0-083047018155d4c00e9977f5692490b3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/hls/2025.05-4da036171b10090d7d21e15c49bb27a1/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/lcg/root/6.32.13-b42babf347314b02960f7ec363565e45/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/tbb/v2022.0.0-fe4e5fc27f2e8eae5da4b66ede97f22b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/cms/vdt/0.4.3-f008bd09702a81e8a062bbef698ed3af/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/cuda/12.9.0-cd4613335deff6e5576167f7dbef4cb0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/json/3.11.3-fe8b14080d05d6ff46ded7c91e3977bc/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/tinyxml2/6.2.0-a0ad3950415fa3138d99b7da42eb4c9f/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_HOST_ONLY -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalCalo/HGCalRecAlgos/plugins/RecoLocalCaloHGCalRecAlgosPluginsPortableCudaAsync/alpaka/HGCalRecHitConfigurationESProducer.cc.d src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitConfigurationESProducer.cc -o tmp/el8_amd64_gcc12/src/RecoLocalCalo/HGCalRecAlgos/plugins/RecoLocalCaloHGCalRecAlgosPluginsPortableCudaAsync/alpaka/HGCalRecHitConfigurationESProducer.cc.o
>> Compiling alpaka/cuda edm plugin src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducers.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-05-23-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-05-23-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_15_1_X_2025-05-23-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/dd4hep/v01-29-00-a6251f4ae4aef60eab61c637c5fb989d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/alpaka/1.2.0-9ea5b7b50bc68454be5c438069f1aa33/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/boost/1.80.0-083047018155d4c00e9977f5692490b3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/hls/2025.05-4da036171b10090d7d21e15c49bb27a1/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/lcg/root/6.32.13-b42babf347314b02960f7ec363565e45/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/tbb/v2022.0.0-fe4e5fc27f2e8eae5da4b66ede97f22b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/cms/vdt/0.4.3-f008bd09702a81e8a062bbef698ed3af/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/cuda/12.9.0-cd4613335deff6e5576167f7dbef4cb0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/json/3.11.3-fe8b14080d05d6ff46ded7c91e3977bc/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02890/el8_amd64_gcc12/external/tinyxml2/6.2.0-a0ad3950415fa3138d99b7da42eb4c9f/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_HOST_ONLY -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalCalo/HGCalRecAlgos/plugins/RecoLocalCaloHGCalRecAlgosPluginsPortableCudaAsync/alpaka/HGCalRecHitProducers.cc.d src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducers.cc -o tmp/el8_amd64_gcc12/src/RecoLocalCalo/HGCalRecAlgos/plugins/RecoLocalCaloHGCalRecAlgosPluginsPortableCudaAsync/alpaka/HGCalRecHitProducers.cc.o
src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationESProducer.cc: In member function 'std::optional<PortableHostCollection<hgcalrechit::HGCalCalibParamSoALayout<> > > alpaka_cuda_async::hgcalrechit::HGCalCalibrationESProducer::produce(const HGCalModuleConfigurationRcd&)':
src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationESProducer.cc:109:18: error: 'LogWarning' is not a member of 'edm'
  109 |             edm::LogWarning("HGCalCalibrationESProducer")
      |                  ^~~~~~~~~~
src/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationESProducer.cc:118:20: error: 'LogWarning' is not a member of 'edm'
  118 |               edm::LogWarning("HGCalCalibrationESProducer")
      |                    ^~~~~~~~~~
</pre>

--- a/CondFormats/HGCalObjects/interface/HGCalMappingModuleIndexer.h
+++ b/CondFormats/HGCalObjects/interface/HGCalMappingModuleIndexer.h
@@ -13,7 +13,6 @@
 #include "CondFormats/HGCalObjects/interface/HGCalDenseIndexerBase.h"
 #include "CondFormats/HGCalObjects/interface/HGCalMappingCellIndexer.h"
 #include "FWCore/Utilities/interface/Exception.h"
-#include "FWCore/MessageLogger/interface/MessageLogger.h"
 
 /**
  * @short this structure holds the indices and types in the readout sequence
@@ -61,21 +60,10 @@ public:
   void finalize();
 
   /**
-   * @short decodes silicon or sipm type and cell type for the detector id 
-   * from the typecode string
+   * @short decode silicon or sipm type and cell type for the detector id 
+   * from the typecode string: "M[LH]-X[123]X-*" for Si, "T[LH]-L*S*[PN]" for SiPm
    */
-  static std::pair<bool, int> convertTypeCode(std::string_view typecode) {
-    if (typecode.size() < 5)
-      throw cms::Exception("InvalidHGCALTypeCode") << typecode << " is invalid for decoding readout cell type";
-    bool isSiPM = {typecode.find("TM") != std::string::npos ? true : false};
-    int celltype;
-    if (isSiPM) {
-      celltype = 0;  // assign SiPM type coarse or molded with next version of modulelocator
-    } else {
-      celltype = {typecode[4] == '1' ? 0 : typecode[4] == '2' ? 1 : 2};
-    }
-    return std::pair<bool, bool>(isSiPM, celltype);
-  }
+  static std::pair<bool, int8_t> getCellType(std::string_view typecode);
 
   /**
    * @short returns the index for the n-th module in the readout sequence of a FED
@@ -136,7 +124,7 @@ public:
   uint32_t getMaxModuleSize() const {
     return maxModulesIdx_;
   }  ///< total number of ECON-Ds (useful for setting ECON-D SoA size)
-  uint32_t getNModules(uint32_t fedid) const {
+  uint32_t getNumModules(uint32_t fedid) const {
     return fedReadoutSequences_[fedid].readoutTypes_.size();
   }  ///< number of ECON-Ds for given FED id
   uint32_t getMaxERxSize() const {
@@ -162,7 +150,7 @@ public:
   }  ///< total number of channels for a given ECON-D typecode
 
   /**
-     @short return type ECON-D Module
+   * @short return type ECON-D Module
    */
   int getTypeForModule(uint32_t fedid, uint32_t modid) const {
     return fedReadoutSequences_[fedid].readoutTypes_[modid];

--- a/CondFormats/HGCalObjects/interface/HGCalMappingParameterSoA.h
+++ b/CondFormats/HGCalObjects/interface/HGCalMappingParameterSoA.h
@@ -19,7 +19,7 @@ namespace hgcal {
                       SOA_COLUMN(int, i1),
                       SOA_COLUMN(int, i2),
                       SOA_COLUMN(uint8_t, irot),
-                      SOA_COLUMN(int, celltype),
+                      SOA_COLUMN(int8_t, celltype),
                       SOA_COLUMN(uint16_t, typeidx),
                       SOA_COLUMN(uint16_t, fedid),
                       SOA_COLUMN(uint16_t, slinkidx),

--- a/CondFormats/HGCalObjects/src/HGCalMappingModuleIndexer.cc
+++ b/CondFormats/HGCalObjects/src/HGCalMappingModuleIndexer.cc
@@ -1,6 +1,14 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "CondFormats/HGCalObjects/interface/HGCalMappingModuleIndexer.h"
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetId.h"       // for HGCSiliconDetId::waferType
+#include "DataFormats/ForwardDetId/interface/HGCScintillatorDetId.h"  // for HGCScintillatorDetId::tileGranularity
 
-//
+/**
+ * @short for a new module it adds it's type to the readaout sequence vector
+ * if the fed id is not yet existing in the mapping it's added
+ * a dense indexer is used to create the necessary indices for the new module
+ * unused indices will be set with -1
+ */
 void HGCalMappingModuleIndexer::processNewModule(uint32_t fedid,
                                                  uint16_t captureblockIdx,
                                                  uint16_t econdIdx,
@@ -48,7 +56,7 @@ void HGCalMappingModuleIndexer::processNewModule(uint32_t fedid,
   }
 }
 
-//
+/// @short to be called after all the modules have been processed
 void HGCalMappingModuleIndexer::finalize() {
   //max indices at different levels
   nfeds_ = fedReadoutSequences_.size();
@@ -121,6 +129,48 @@ void HGCalMappingModuleIndexer::finalize() {
       typeCounters[type_val]++;
     }
   }
+}
+
+/**
+ * @short decode silicon or sipm type and cell type for the detector id 
+ * from the typecode string: "M[LH]-X[123]X-*" for Si, "T[LH]-L*S*[PN]" for SiPm
+ */
+std::pair<bool, int8_t> HGCalMappingModuleIndexer::getCellType(std::string_view typecode) {
+  if (typecode.size() < 5) {
+    cms::Exception ex("InvalidHGCALTypeCode");
+    ex << "'" << typecode << "' is invalid for decoding readout cell type";
+    ex.addContext("Calling HGCalMappingModuleIndexer::getCellType()");
+    throw ex;
+  }
+  int8_t celltype = -1;
+  const bool isSiPM = (typecode[0] == 'T');
+  const bool isHD = (typecode[1] == 'H');
+  if (isSiPM) {  // assign SiPM type coarse or molded with next version of modulelocator
+    if (isHD)
+      celltype = HGCScintillatorDetId::tileGranularity::HGCalTileFine;
+    else
+      celltype = HGCScintillatorDetId::tileGranularity::HGCalTileNormal;
+  } else {  // assign Si wafer type low/high density and thickness (120, 200, 300 um)
+    const char thickness = typecode[4];
+    if (isHD) {
+      if (thickness == '1')
+        celltype = HGCSiliconDetId::waferType::HGCalHD120;
+      else if (thickness == '2')
+        celltype = HGCSiliconDetId::waferType::HGCalHD200;
+    } else {
+      if (thickness == '2')
+        celltype = HGCSiliconDetId::waferType::HGCalLD200;
+      else if (thickness == '3')
+        celltype = HGCSiliconDetId::waferType::HGCalLD300;
+    }
+  }
+  if (celltype == -1) {
+    cms::Exception ex("InvalidHGCALTypeCode");
+    ex << "Could not parse cell type from typecode='" << typecode << "'";
+    ex.addContext("Calling HGCalMappingModuleIndexer::getCellType()");
+    throw ex;
+  }
+  return std::pair<bool, int8_t>(isSiPM, celltype);
 }
 
 //

--- a/Geometry/HGCalMapping/plugins/alpaka/HGCalMappingModuleESProducer.cc
+++ b/Geometry/HGCalMapping/plugins/alpaka/HGCalMappingModuleESProducer.cc
@@ -64,7 +64,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
           int typeidx = modIndexer.getTypeForModule(fedid, captureblockidx, econdidx);
           std::string typecode = pmap.getAttr("typecode", row);
 
-          auto celltypes = modIndexer.convertTypeCode(typecode);
+          auto celltypes = modIndexer.getCellType(typecode);
           bool isSiPM = celltypes.first;
           int celltype = celltypes.second;
           int zside = pmap.getIntAttr("zside", row);

--- a/RecoLocalCalo/HGCalRecAlgos/src/HGCalESProducerTools.cc
+++ b/RecoLocalCalo/HGCalRecAlgos/src/HGCalESProducerTools.cc
@@ -28,7 +28,6 @@ namespace hgcal {
        << data.begin().key() << "'...";
     ex.addContext("Calling hgcal::search_modkey()");
     throw ex;
-    return data.begin().key();  // no matching key found in whole JSON map
   }
 
   // @short search first match to a given FED index in a JSON (following insertion order)
@@ -74,6 +73,7 @@ namespace hgcal {
       ex << "Could not find matching key for '" << fedid << "' in '" << name << "'! Returning first key '" << matchedkey
          << "'...";
       ex.addContext("Calling hgcal::search_fedkey()");
+      throw ex;
     } else {
       edm::LogInfo("search_fedkey") << "search_fedkey: Matched module='" << fedid << "' to fedkey='" << matchedkey
                                     << "'";

